File: plugin/ios/Plugin/Translator.swift
import Foundation

/// High-level translation API
/// Coordinates tokenizer, engine, and decoder
class Translator {
    
    private let tokenizer: SimpleBPETokenizer
    private let engine: TranslationEngine
    private let decoder: BeamSearchDecoder
    private let languageTokenMap: [String: Int]
    
    private init(tokenizer: SimpleBPETokenizer, engine: TranslationEngine,
                 decoder: BeamSearchDecoder, languageTokenMap: [String: Int]) {
        self.tokenizer = tokenizer
        self.engine = engine
        self.decoder = decoder
        self.languageTokenMap = languageTokenMap
    }
    
    /// Create Translator from model directory
    static func create(modelsDir: URL) throws -> Translator {
        // Load tokenizer
        let tokenizer = try loadTokenizer(from: modelsDir)
        print("[Translator] Tokenizer loaded, vocab size: \(tokenizer.vocabSize)")
        
        // Load language tokens
        let langTokens = try loadLanguageTokens(from: modelsDir)
        print("[Translator] Language tokens loaded: \(langTokens.count)")
        
        // Create engine and load models
        let engine = try TranslationEngine()
        try engine.loadModels(from: modelsDir)
        
        // Create beam search decoder
        let decoder = BeamSearchDecoder(eosTokenId: tokenizer.eosTokenId)
        
        return Translator(tokenizer: tokenizer, engine: engine, decoder: decoder, languageTokenMap: langTokens)
    }
    
    /// Translate text to target language
    func translate(text: String, targetLanguage: String) throws -> String {
        print("[Translator] Translating: \"\(text)\" to \(targetLanguage)")
        
        // Get target language token ID
        let tgtToken = "__\(targetLanguage)__"
        guard let tgtTokenId = languageTokenMap[tgtToken] else {
            throw TranslationError.unknownLanguage(targetLanguage)
        }
        
        // Tokenize input
        let textTokens = tokenizer.encode(text)
        
        // Build encoder input: [tgt_lang_token, ...tokens, eos_token]
        var inputIds: [Int64] = [Int64(tgtTokenId)]
        inputIds.append(contentsOf: textTokens.map { Int64($0) })
        inputIds.append(Int64(tokenizer.eosTokenId))
        
        // Attention mask (all 1s)
        let attentionMask = [Int64](repeating: 1, count: inputIds.count)
        
        // Run encoder
        let startEncoder = Date()
        try engine.runEncoder(inputIds: inputIds, attentionMask: attentionMask)
        print("[Translator] Encoder: \(Int(Date().timeIntervalSince(startEncoder) * 1000))ms")
        
        // Beam search decode
        let startDecode = Date()
        let startTokens = [tokenizer.eosTokenId]
        
        let outputIds = try decoder.decode(startTokenIds: startTokens, maxNewTokens: 256) { ids in
            try self.engine.runDecoderStep(decoderInputIds: ids)
        }
        print("[Translator] Decoder: \(Int(Date().timeIntervalSince(startDecode) * 1000))ms")
        
        // Clear engine cache
        engine.clearCache()
        
        // Detokenize result
        let result = detokenize(outputIds)
        print("[Translator] Result: \"\(result)\"")
        
        return result
    }
    
    /// Check if translator is ready
    var isReady: Bool {
        return engine.isReady
    }
    
    /// Release resources
    func close() {
        engine.close()
    }
    
    /// Get supported language codes
    var supportedLanguages: [String] {
        languageTokenMap.keys.map { $0.replacingOccurrences(of: "__", with: "") }
    }
    
    // For debug
    var translationEngine: TranslationEngine { engine }
    
    // MARK: - Private
    
    private func detokenize(_ ids: [Int]) -> String {
        guard ids.count > 1 else { return "" }
        
        var outputIds = Array(ids.dropFirst())
        
        // Remove EOS at end if present
        if let last = outputIds.last, last == tokenizer.eosTokenId {
            outputIds.removeLast()
        }
        
        // Filter special and language tokens
        outputIds = outputIds.filter { id in
            id != tokenizer.eosTokenId &&
            id != tokenizer.bosTokenId &&
            id != tokenizer.padTokenId &&
            id < 128000
        }
        
        return tokenizer.decode(outputIds, skipSpecialTokens: true)
    }
    
    private static func loadTokenizer(from modelsDir: URL) throws -> SimpleBPETokenizer {
        let vocabURL = modelsDir.appendingPathComponent("vocab.json")
        let data = try Data(contentsOf: vocabURL)
        guard let vocab = try JSONSerialization.jsonObject(with: data) as? [String: Int] else {
            throw NSError(domain: "Translator", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid vocab.json"])
        }
        return SimpleBPETokenizer(vocab: vocab)
    }
    
    private static func loadLanguageTokens(from modelsDir: URL) throws -> [String: Int] {
        let url = modelsDir.appendingPathComponent("added_tokens.json")
        let data = try Data(contentsOf: url)
        guard let tokens = try JSONSerialization.jsonObject(with: data) as? [String: Int] else {
            throw NSError(domain: "Translator", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid added_tokens.json"])
        }
        return tokens
    }
}

File: plugin/android/src/main/java/com/small100onnx/Translator.java
package com.small100onnx;

import android.util.Log;

import org.json.JSONObject;

import java.io.File;
import java.io.FileInputStream;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Translator - high-level translation API
 * Coordinates tokenizer, engine, and decoder
 */
public class Translator {
    private static final String TAG = "Translator";
    
    private final SimpleBPETokenizer tokenizer;
    private final TranslationEngine engine;
    private final BeamSearchDecoder decoder;
    private final Map<String, Integer> languageTokenMap;
    
    private Translator(SimpleBPETokenizer tokenizer, TranslationEngine engine,
                       BeamSearchDecoder decoder, Map<String, Integer> languageTokenMap) {
        this.tokenizer = tokenizer;
        this.engine = engine;
        this.decoder = decoder;
        this.languageTokenMap = languageTokenMap;
    }
    
    /**
     * Create Translator from model directory
     */
    public static Translator create(File modelsDir) throws Exception {
        // Load tokenizer
        SimpleBPETokenizer tokenizer = loadTokenizer(modelsDir);
        Log.d(TAG, "Tokenizer loaded, vocab size: " + tokenizer.getVocabSize());
        
        // Load language tokens
        Map<String, Integer> langTokens = loadLanguageTokens(modelsDir);
        Log.d(TAG, "Language tokens loaded: " + langTokens.size());
        
        // Create engine and load models
        TranslationEngine engine = new TranslationEngine();
        engine.loadModels(modelsDir);
        
        // Create beam search decoder
        BeamSearchDecoder decoder = new BeamSearchDecoder(tokenizer.getEosTokenId());
        
        return new Translator(tokenizer, engine, decoder, langTokens);
    }
    
    /**
     * Translate text to target language
     */
    public String translate(String text, String targetLanguage) throws Exception {
        Log.d(TAG, "Translating: \"" + text + "\" to " + targetLanguage);
        
        // Get target language token ID
        String tgtToken = "__" + targetLanguage + "__";
        Integer tgtTokenId = languageTokenMap.get(tgtToken);
        if (tgtTokenId == null) {
            throw new IllegalArgumentException("Unknown target language: " + targetLanguage);
        }
        
        // Tokenize input
        int[] textTokens = tokenizer.encode(text);
        
        // Build encoder input: [tgt_lang_token, ...tokens, eos_token]
        long[] inputIds = new long[textTokens.length + 2];
        inputIds[0] = tgtTokenId;
        for (int i = 0; i < textTokens.length; i++) {
            inputIds[i + 1] = textTokens[i];
        }
        inputIds[inputIds.length - 1] = tokenizer.getEosTokenId();
        
        // Attention mask (all 1s)
        long[] attentionMask = new long[inputIds.length];
        Arrays.fill(attentionMask, 1L);
        
        // Run encoder
        long startEncoder = System.currentTimeMillis();
        engine.runEncoder(inputIds, attentionMask);
        Log.d(TAG, "Encoder: " + (System.currentTimeMillis() - startEncoder) + "ms");
        
        // Beam search decode
        long startDecode = System.currentTimeMillis();
        int[] startTokens = {tokenizer.getEosTokenId()};
        
        int[] outputIds = decoder.decode(startTokens, 256, ids -> engine.runDecoderStep(ids));
        Log.d(TAG, "Decoder: " + (System.currentTimeMillis() - startDecode) + "ms");
        
        // Clear engine cache
        engine.clearCache();
        
        // Detokenize result
        String result = detokenize(outputIds);
        Log.d(TAG, "Result: \"" + result + "\"");
        
        return result;
    }
    
    /**
     * Check if translator is ready
     */
    public boolean isReady() {
        return engine.isReady();
    }
    
    /**
     * Release resources
     */
    public void close() {
        engine.close();
    }
    
    /**
     * Get supported language codes
     */
    public String[] getSupportedLanguages() {
        return languageTokenMap.keySet().stream()
            .map(s -> s.replace("__", ""))
            .toArray(String[]::new);
    }
    
    // For debug
    public TranslationEngine getEngine() {
        return engine;
    }
    
    // Private helpers
    
    private String detokenize(int[] ids) {
        // Skip first token (EOS start) and last EOS
        if (ids.length <= 1) return "";
        
        int end = ids.length;
        if (ids[end - 1] == tokenizer.getEosTokenId()) end--;
        
        // Filter special and language tokens
        int[] filtered = Arrays.stream(ids, 1, end)
            .filter(id -> id != tokenizer.getEosTokenId() 
                       && id != tokenizer.getBosTokenId()
                       && id != tokenizer.getPadTokenId()
                       && id < 128000)
            .toArray();
        
        return tokenizer.decode(filtered, true);
    }
    
    private static SimpleBPETokenizer loadTokenizer(File modelsDir) throws Exception {
        File vocabFile = new File(modelsDir, "vocab.json");
        String json = readFile(vocabFile);
        JSONObject vocabObj = new JSONObject(json);
        
        Map<String, Integer> vocab = new HashMap<>();
        Iterator<String> keys = vocabObj.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            vocab.put(key, vocabObj.getInt(key));
        }
        
        return new SimpleBPETokenizer(vocab);
    }
    
    private static Map<String, Integer> loadLanguageTokens(File modelsDir) throws Exception {
        File file = new File(modelsDir, "added_tokens.json");
        String json = readFile(file);
        JSONObject obj = new JSONObject(json);
        
        Map<String, Integer> result = new HashMap<>();
        Iterator<String> keys = obj.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            result.put(key, obj.getInt(key));
        }
        
        return result;
    }
    
    private static String readFile(File file) throws Exception {
        try (FileInputStream fis = new FileInputStream(file)) {
            byte[] data = new byte[(int) file.length()];
            fis.read(data);
            return new String(data, StandardCharsets.UTF_8);
        }
    }
}
